src/steamapi.rs

```diff
@@ -16,33 +19,58 @@ use crate::{
     state::State,
 };
 
+const BATCH_INTERVAL: Duration = Duration::from_millis(500);
+const BATCH_SIZE: usize = 20;  // adjust as needed
+
 /// Enter a loop to wait for steam lookup requests, make those requests from the Steam web API,
 /// and update the state to include that data. Intended to be run inside a new tokio::task
 pub async fn steam_api_loop(mut requests: UnboundedReceiver<SteamID>, api_key: Arc<str>) {
     tracing::debug!("Entering steam api request loop");
 
     let mut client = SteamAPI::new(api_key);
+    let mut buffer: VecDeque<SteamID> = VecDeque::new();
+    let mut batch_timer = tokio::time::interval(BATCH_INTERVAL);
+
     loop {
-        if let Some(request) = requests.recv().await {
-            match request_steam_info(&mut client, request).await {
-                Ok(steam_info) => {
-                    State::write_state()
-                        .server
-                        .insert_steam_info(request, steam_info);
+        tokio::select! {
+            Some(request) = requests.recv() => {
+                buffer.push_back(request);
+                if buffer.len() >= BATCH_SIZE {
+                    send_batch(&mut client, &mut buffer).await;
+                    batch_timer.tick().await;  // Reset the timer
```

Comment:
I don't 100% understand how tokio intervals work, but I don't think this is the right way to be going about this. The [tokio::interval documentation](https://docs.rs/tokio/latest/tokio/time/fn.interval.html) says the default `MissedTickBehaviour` is [Burst](https://docs.rs/tokio/latest/tokio/time/enum.MissedTickBehavior.html#variant.Burst), which means any missed ticks will be caught up by immediately returning from tick, which I imagine is what you're trying to overcome by calling `batch_time.tick()` here, but I think this can be avoided completely by setting the `MissedTickBehaviour` to [Delay](https://docs.rs/tokio/latest/tokio/time/enum.MissedTickBehavior.html#variant.Delay).

Even with `Delay` though, I think in the case that we send off a batch of requests and it returns before the next tick is due, next time `tick()` is called it will return sooner because we never actually missed a tick. I'm not 100% sure on the behaviour but that is what I'm understanding from the weirdly-worded and insufficient documentation. But that makes me think that `Interval` might not be the right tool for the job as opposed to just trying to receive the next request inside a [timeout](https://docs.rs/tokio/latest/tokio/time/fn.timeout.html).

Reply:
Actually a `timeout` will give differing behaviour still. I think if you change the `MissedTickBehaviour` to `Delay`, and then change line 40 to call `batch_timer.reset()` instead that should be correct.